plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +232,152 @@ class bp_connection_manager {
+      if (!initial_msg) {
+         // validate structure and data of msg
+         auto valid_address = [](const std::string& addr) -> bool {
+            const auto& [host, port, type] = net_utils::split_host_port_type(addr);
+            return !host.empty() && !port.empty();
+         };
+         const gossip_bp_peers_message::bp_peer* prev = nullptr;
+         size_t num_per_producer = 0;
+         for (const auto& peer : msg.peers) {
+            if (peer.producer_name.empty())
+               return false; // invalid bp_peer data
+            if (!valid_address(peer.server_address))
+               return false; // invalid address
+            if (prev != nullptr) {
+               if (prev->producer_name == peer.producer_name) {
+                  ++num_per_producer;
+                  if (num_per_producer > max_bp_peers_per_producer)
+                     return false; // more than allowed per producer
+                  if (prev->server_address == peer.server_address)
+                     return false; // duplicate entries not allowed
+               } else if (prev->producer_name > peer.producer_name) {
+                  return false; // required to be sorted
+               } else {
+                  num_per_producer = 0;
+               }
+            }
+            prev = &peer;
+         }
+      }
+
+      const controller& cc = self()->chain_plug->chain();
+      bool fatal_error = false;
+      auto is_peer_key_valid = [&](const gossip_bp_peers_message::bp_peer& peer) -> bool {
+         try {
+            if (peer.sig.is_webauthn()) {
+               fc_dlog(self()->get_logger(), "Peer ${p} signature is webauthn, not allowed.", ("p", peer.producer_name));
+               fatal_error = true;
+               return false;
+            }
+            std::optional<peer_info_t> peer_info = cc.get_peer_info(peer.producer_name);
+            if (peer_info && peer_info->key) {
+               constexpr bool check_canonical = false;
+               public_key_type pk(peer.sig, peer.digest(), check_canonical);
+               if (pk != *peer_info->key) {
+                  fc_dlog(self()->get_logger(), "Recovered peer key did not match on-chain ${p}, recovered: ${pk} != expected: ${k}",
+                          ("p", peer.producer_name)("pk", pk)("k", *peer_info->key));
+                  return false;
+               }
+            } else { // unknown key
+               // ok, might have just been deleted or dropped out of top ranking
+               fc_dlog(self()->get_logger(), "Failed to find peer key ${p}", ("p", peer.producer_name));
+               return false;
+            }
+         } catch (fc::exception& e) {
+            fc_dlog(self()->get_logger(), "Exception recovering peer key ${p}, error: ${e}", ("p", peer.producer_name)("e", e.to_detail_string()));
+            fatal_error = true;
+            return false; // invalid key
+         }
```

Comment:
Can something throw an exception in this lambda?
If not we can remove this `fatal_error` stuff.

Reply:
Yes. `public_key_type` constructor can throw on invalid format for example.